numpy/_core/src/multiarray/ctors.c

```diff
@@ -510,14 +512,17 @@ PyArray_AssignFromCache_Recursive(
     }
     else {
         assert(depth != ndim);
+        // this macro thakes *the argument* of PySequence_Fast, which is orig_obj;
+        // not the object returned by PySequence_Fast, which is obj
+        NPY_BEGIN_CRITICAL_SECTION_SEQUENCE_FAST(orig_obj);
```

Comment:
This can still crash and produce data races if the sequence gets mutated but the size remains same. Inside the loop, the items of sequences are accessed using borrowed references (`PySequence_Fast_GET_ITEM`) so if the critical section gets suspended and other thread sets a new object at current index then it can crash with use-after-free. 

You should be able to reproduce this case with a test which tries to create an `ndarray` from a list whose size is fixed but it getting concurrently modified by other thread with `list.__setitem__`.

Reply:
I guess I'm trying to ask: how would you suggest fixing the issue?

Reply:
You can try to further reduce the critical section and check the size after each iteration, so something like:
```
for (int i = 0; i = original_length; i++) {
NPY_BEGIN_CRITICAL_SECTION_SEQUENCE_FAST(orig_obj);
if (check length changed) {
    raise size changed exception
}
PyObject *value = PySequence_Fast_GET_ITEM(obj, i);
INCREF(value);
END_CRITICAL_SECTION();
// rest of loop
}
```

The strong reference is still needed though but I think with the reduced critical section, it would not cause a lot of contention.